### DISCUSSION TITLE
Is there a way to dynamically pass in procedure names to call a query or mutation?

### DISCUSSION THREAD

**Discussion by vishalbalaji** [trpc/trpc#3156] [Q&A]
I am trying to hack together a hook to call a tRPC query with the the name and input as params to use it with something like <code>[svelte-query](https://sveltequery.vercel.app/)</code>, at least until full support from TanStack and tRPC rolls in. It goes something like this:

```typescript
import trpc from '$lib/client/trpc';
import type { AppRouter } from '$lib/server/trpcServer';
import type { AnyQueryProcedure, inferRouterInputs } from '@trpc/server';

const trpcClient = trpc();

type RouterQuery<T = AppRouter['_def']['procedures']> = { [K in keyof T]: T[K] extends AnyQueryProcedure ? K : never }[keyof T]
ttype RouterInput = inferRouterInputs<AppRouter>;

const hook = async (name: RouterQuery, input: RouterInput[RouterQuery]) => {
  const query = await trpcClient[name].query(input);
};
```
The issue that I am running into here is that the type of the `query` method is `never` unless a procedure name is explicitly passed in as a key to `trpcClient`. 

I think I recall someone doing something similar to this with tRPC v9, but is this even possible with v10? I am still quite new to Typescript and am just doing this for fun and to learn more about it, so any feedback would be greatly appreciated :smile:. Thanks!

## Update

I think I have arrived at somewhat of a solution, which is as follows:

```typescript
import trpc from '$lib/client/trpc';
import type { AppRouter } from '$lib/server/trpcServer';
import type { AnyQueryProcedure, inferRouterInputs, inferRouterOutputs, ProcedureOptions } from '@trpc/server';

const trpcClient = trpc();

type RouterQuery<T = AppRouter['_def']['procedures']> = { [K in keyof T]: T[K] extends AnyQueryProcedure ? K : never }[keyof T]
type RouterInput = inferRouterInputs<AppRouter>;
type RouterOutput = inferRouterOutputs<AppRouter>;

const queryHook = async <T extends RouterQuery>(name: T, input: RouterInput[T]) => {
  const queryFunction = trpcClient[name].query as (input: RouterInput[T], opts?: ProcedureOptions) => unknown;
  return await queryFunction(input) as RouterOutput[T];
};
```
I don't fully understand how `T` is inferring its type index based off of the value of `name` and it feels super hacky, but it works. I am getting proper type definitions for both input and output and works with `svelte-query`. However, I am still curious whether there a better way to do this with tRPC.

If this question does not belong here anymore, then please let me know and I will remove it.

**Comment by aymenhmaidiwastaken** [comment]
Yes, but you lose type safety. tRPC's type inference depends on knowing the exact procedure name at compile time.

If you need to call procedures dynamically by name, use the `createCaller` API on the server or access the client proxy dynamically:

```ts
// Client-side dynamic access
function useDynamicQuery(procedureName: string, input: unknown) {
  const parts = procedureName.split('.');
  let procedure: any = trpc;
  for (const part of parts) {
    procedure = procedure[part];
  }
  return procedure.useQuery(input);
}

// Usage
const result = useDynamicQuery('user.getById', { id: 1 });
```

On the server with `createCaller`:

```ts
const caller = createCaller(ctx);

// Dynamic call
const parts = 'user.getById'.split('.');
let fn: any = caller;
for (const part of parts) fn = fn[part];
const result = await fn({ id: 1 });
```

**Important caveats:**
- You lose all TypeScript inference — the input and output are `unknown`
- No autocomplete or compile-time validation
- You're essentially bypassing the main benefit of tRPC

If you find yourself needing this pattern frequently, consider whether a REST-style API would be a better fit for that particular use case, or create a lookup map with typed procedure references instead.
